src/lib.rs

```diff
@@ -73,22 +74,67 @@ use std::io::Bytes;
 use std::io::Read;
 use std::io::Write;
 use std::iter::Peekable;
+use std::num::ParseIntError;
 use std::ops::Deref;
+use thiserror::Error;
 
 /////////////////////
 
 /// The error type for reading and writing properties files.
-#[derive(Debug)]
+#[derive(Debug, Error)]
+#[error("{cause} (line_number = {})", self.display_line_number())]
 pub struct PropertiesError {
-  description: String,
-  cause: Option<Box<dyn Error + 'static + Send + Sync>>,
+  #[source]
+  cause: PropertiesErrorCause,
   line_number: Option<usize>,
 }
 
+/// Underlying errors for [`PropertiesError`]
+#[non_exhaustive]
+#[derive(Debug, Error)]
+pub enum PropertiesErrorCause {
```

Comment:
It doesn't make sense for this struct to implement `Error` as well.  Having separate `IOError`, `EncodingError`, etc. cases doesn't provide much benefit, because the user could always just call `source()` and downcast.

It probably makes more sense to create `BadKeyValueSeparator` and `BadCommentPrefix` structs, implement `Error` for them, make all of these direct causes of `PropertiesError`, and remove the `PropertiesErrorCause` struct.